core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ProfileCredentialsProvider.java

```diff
@@ -127,19 +115,45 @@ public AwsCredentials resolveCredentials() {
         if (loadException != null) {
             throw loadException;
         }
+
+        ProfileFile cachedOrRefreshedProfileFile = refreshProfileFile();
+        if (isNewProfileFile(cachedOrRefreshedProfileFile)) {
+            synchronized (this) {
+                if (isNewProfileFile(cachedOrRefreshedProfileFile)) {
+                    currentProfileFile = cachedOrRefreshedProfileFile;
+                    handleProfileFileReload(cachedOrRefreshedProfileFile);
+                }
+            }
+        }
+
         return credentialsProvider.resolveCredentials();
     }
 
+    private void handleProfileFileReload(ProfileFile profileFile) {
+        credentialsProvider = createCredentialsProvider(profileFile, profileName);
+    }
+
+    private ProfileFile refreshProfileFile() {
+        return profileFile.get();
+    }
+
+    private boolean isNewProfileFile(ProfileFile profileFile) {
+        return !Objects.equals(currentProfileFile, profileFile);
+    }
+
     @Override
     public String toString() {
         return ToString.builder("ProfileCredentialsProvider")
                        .add("profileName", profileName)
-                       .add("profileFile", profileFile)
+                       .add("profileFile", currentProfileFile)
                        .build();
     }
 
     @Override
     public void close() {
+        if (profileFile instanceof SdkAutoCloseable) {
+            ((SdkAutoCloseable) profileFile).close();
+        }
```

Comment:
This is a red flag to me: closing something we didn't necessarily create. What if this is being used by someone else and we just closed it/broke it?

My bigger question is: why are profile file suppliers closeable?